netdata/netdata
# eBPF Functions (enable/disable threads)

##### Summary
This is the first PR bringing functions to eBPF.plugin;

##### Test Plan

1. Copy `tests/ebpf/ebpf.d.conf` to `/etc/netdata/`
2. Compile this branch with flag `NETDATA_DEV_MODE`.
3. Run next command and verify that everything succeed:

```sh
$ sh tests/ebpf/ebpf_thread_function.sh > out.txt 2> err.tx 
```

##### Additional Information
This PR was tested on :

| Distribution | Environment | Kernel | Error | Output |
|-------------------|---------------------|-----------|------------|---------|
|Slackware current | Bare metal | 6.1.36 | [slackware_6_1_err.txt](https://github.com/netdata/netdata/files/11941169/slackware_6_1_err.txt) | [slackware_6_1_out.txt](https://github.com/netdata/netdata/files/11941170/slackware_6_1_out.txt) |
| Arch Linux | Libvirt | 6.4.1-arch1-1|[arch_6_4_err.txt](https://github.com/netdata/netdata/files/11941704/arch_6_4_err.txt) | [arch_6_4_out.txt](https://github.com/netdata/netdata/files/11941705/arch_6_4_out.txt) |
|Ubuntu 22.04 | Libvirt | 5.15.0-69-generic |  [ubuntu_5_15_err.txt](https://github.com/netdata/netdata/files/11941849/ubuntu_5_15_err.txt) | [ubuntu_5_15_out.txt](https://github.com/netdata/netdata/files/11941850/ubuntu_5_15_out.txt) | 
|CentOS 9 | Libvirt | 5.14.0-331.el9.x86_64  | [centos_5_14_err.txt](https://github.com/netdata/netdata/files/11943347/centos_5_14_err.txt) | [centos_5_14_out.txt](https://github.com/netdata/netdata/files/11943348/centos_5_14_out.txt) |
| Debian 11 | Libvirt | 5.10.0-23-amd64 | [debian_5_10_err.txt](https://github.com/netdata/netdata/files/11949024/debian_5_10_err.txt) | [debian_5_10_out.txt](https://github.com/netdata/netdata/files/11949025/debian_5_10_out.txt) | 
|Ubuntu 20.04 | Libvirt | 5.4.0-146-generic | [ubuntu_5_4_err.txt](https://github.com/netdata/netdata/files/11896306/ubuntu_5_4_err.txt) | [ubuntu_5_4_out.txt](https://github.com/netdata/netdata/files/11896307/ubuntu_5_4_out.txt) | 
| Alma 8 | Libvirt | 4.18.0-477.13.1.el8_8.x86_64 | [alma_4_18_err.txt](https://github.com/netdata/netdata/files/11951013/alma_4_18_err.txt) | [alma_4_18_out.txt](https://github.com/netdata/netdata/files/11951014/alma_4_18_out.txt) |
| Ubuntu 18.04 | Libvirt | 4.15.0-208-generic | [ubuntu_4_15_err.txt](https://github.com/netdata/netdata/files/11951655/ubuntu_4_15_err.txt) | [ubuntu_4_15_out.txt](https://github.com/netdata/netdata/files/11951656/ubuntu_4_15_out.txt) | 
|Slackware current | Qemu| 4.14.290  | [slackware_4_14_err.txt](https://github.com/netdata/netdata/files/11952318/slackware_4_14_err.txt) | [slackware_4_14_out.txt](https://github.com/netdata/netdata/files/11952319/slackware_4_14_out.txt) | 
| CentOS 7 | Libvirt | 3.10.0-1160.92.1.el7.x86_64 | [centos_3_10_err.txt](https://github.com/netdata/netdata/files/11952053/centos_3_10_err.txt) | [centos_3_10_out.txt](https://github.com/netdata/netdata/files/11952054/centos_3_10_out.txt) |

When you run the script it is expected to have charts like this:

![1](https://github.com/netdata/netdata/assets/49162938/91823573-114c-4c16-b634-cc46f7bb1bcf)


<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? eBPF
- Can they see the change or is it an under the hood? If they can see it, where? no, this is an internal controller accessed via API. We are also expecting a cloud schema.
- How is the user impacted by the change?  This PR has potential to reduce overhead.
- What are there any benefits of the change?  A light plugin
</details>
